### PR TITLE
Add `throws` documentation to Modules

### PR DESCRIPTION
**Pull request by piyush0411**
Started working on #11417 
This is my first contribution to Chapel. Kindly review the change.

I shall push more commits to this pull request.

Thanks :)

### CODE REVIEW

File: modules/standard/Regexp.chpl
Comment (lydia-duncan) [inline modules/standard/Regexp.chpl:457]: You'll want to include this as part of the previous comment - putting it here will prevent the other documentation from being include (and it's useful :) )

File: modules/standard/Regexp.chpl
Comment (lydia-duncan) [inline modules/standard/Regexp.chpl:456]: Unfortunately, this is inaccurate - compilerError is a special function that causes compilation to fail.  It's not something that can be caught by a try/catch block.  You want to look in the body for `throw`.  Looking at the function, it looks like we throw a BadRegexpError.  What can you tell about the circumstances under which that happens?

You'll also want to match the style of other `:throws:` documentation.  Take a look at the documentation comment for modules/standard/Random.chpl's `getNth` for an example of how it should be written.

File: modules/standard/Spawn.chpl
Comment (lydia-duncan) [inline modules/standard/Spawn.chpl:260]: Note that this function is marked with `pragma "no doc"`.  That means that it will not turn up in the documentation normally, so doesn't need to have a documentation comment with `:throws:` in it.

**Review by lydia-duncan — COMMENTED** [review]
Hi!  Welcome to Chapel and thanks for taking this effort on!  It looks like you've tripped over a couple of things that I had forgotten weren't easily apparent, sorry for not leaving better guidance on the issue.  Hopefully this helps, please don't be discouraged!

File: modules/standard/Spawn.chpl
Comment (lydia-duncan) [reply modules/standard/Spawn.chpl:260]: I hadn't had time on the issue to go through an eliminate false positives.  Hopefully this helps avoid tripping over them in the future!

File: modules/standard/Regexp.chpl
Comment (piyush0411) [reply modules/standard/Regexp.chpl:456]: The error is thrown when the variable `CHPL_REGEXP` is set to `none`. The introduction to this file tells that regular expression support is built on the top of the RE2 regular expression library and should be initialized with `re2`. So if its value is `none`, the compiler should throw an error. This is what I have understood after going through the file.

I shall change the error type to BadRegexError as you mentioned.

Thanks for the prompt reply :)

File: modules/standard/Spawn.chpl
Comment (piyush0411) [reply modules/standard/Spawn.chpl:260]: I was having this confusion while going through the file and was about to ask you about it.
I'll remove this part.

File: modules/standard/Regexp.chpl
Comment (lydia-duncan) [inline modules/standard/Regexp.chpl:453]: Syntactical feedback: it should be `:throws BadRegexError:`, not `:throws: BadRegexError:`, you have an extra semi colon.

Substantive feedback: again, the compiler error does not throw, it is just a compilation error.  `CHPL_REGEXP` being set to none is what triggers the *compiler* error, not what triggers the `throw` later in the body (on line 492).  The condition that causes `BadRegexpError` to be thrown is the if statement on line 483.  To determine what causes *that* function to return false, you'll need to look at where it is defined (my guess is somewhere in the runtime).

Think of a compiler error as equivalent in severity to failing to include a semi colon or making a typo when trying to call a function - when something like that happens, the compiler doesn't have the tools to adjust for it and needs the person writing the program to make a change and try to compile again.  In this case, having CHPL_REGEXP set to none means the compiler isn't able to include the underlying implementation that supports the Regexp module at all.  This is in contrast to error handling constructs, where something has gone wrong but the user could potentially have written their program to detect the error and adjust as a result.

If you're still confused about error handling, I recommend checking out the [error handling primer](https://chapel-lang.org/docs/latest/primers/errorHandling.html).  You can also read the [chapter in the spec](https://chapel-lang.org/docs/latest/language/spec/error-handling.html).

Comment (piyush0411) [reply modules/standard/Regexp.chpl:453]: Hi @lydia-duncan Mam
I have read the [qio_regexp.h](https://github.com/chapel-lang/chapel/blob/master/runtime/include/qio/qio_regexp.h) file in `runtime` folder. Kindly refer to line [59](https://github.com/chapel-lang/chapel/blob/master/runtime/include/qio/qio_regexp.h#L59) of this file. Here it is mentioned that the function will `Return true for ok` and if false will return an error string. However, I am still not able to find the definition of this function and when will it return false.
So how should the comments look? Can you please help me with this?

Comment (lydia-duncan) [reply modules/standard/Regexp.chpl:453]: I think the function definition you're looking for is [here](https://github.com/chapel-lang/chapel/blob/master/runtime/src/qio/regexp/re2/re2-interface.cc#L203)

I found that file by looking in the `runtime/src` directory.  Since the .h file was under `qio` in the `include` directory, I looked in the `runtime/src/qio` directory, saw a directory that was labelled the same as the module we're looking at, and then from there I made a guess based on knowing that our Regexp implementation is supported by the package `re2` under the covers (which is definitely a leap that would be hard for a new person, sorry).   But I suspect a `grep` command in the runtime directory would have gotten to it faster.

You may have to do some more chasing - I think the main function that function relies upon is still in the same file, but it may be you'll have to look at something `re2` defines to fully determine the conditions.

Comment (piyush0411) [reply modules/standard/Regexp.chpl:453]: I have read the modules for `re2` and `regexp`, studied the function definition of the function we are working with. 
An error is thrown if `re2 && re2->ok()` evaluates to `false`. In [this](https://github.com/chapel-lang/chapel/blob/master/third-party/re2/re2-src/re2/re2.cc) file, in line [1022](https://github.com/chapel-lang/chapel/blob/master/third-party/re2/re2-src/re2/re2.cc#L1022), the error statement is `Invalid RE2`. So should this be the reason for throwing an error?
In my opinion, this error is related to the improper creation of `re2`, as mentioned [here](https://github.com/chapel-lang/chapel/blob/master/third-party/re2/re2-src/re2/re2.h#L286)

Comment (lydia-duncan) [reply modules/standard/Regexp.chpl:453]: I think that's a reasonable interpretation of the code, yeah!

**Comment by lydia-duncan** [comment]
Ack, sorry, it looks like we had another miscommunication here.  I'm going to mark this as post-release but that doesn't mean I'm going to ignore you (just that we won't merge it immediately).  Sorry about that, we'll figure this out and thanks for your patience!  (I'll do a more detailed update tomorrow, I have another review that needs my attention as well)

File: modules/standard/Regexp.chpl
Comment (lydia-duncan) [inline modules/standard/Regexp.chpl:453]: So I interpreted your earlier comment (`re2 && re2->ok()`) as referring to the particular pattern being evaluated rather than the package as a whole, since `->` indicates a method call.  But here, you're referring to RE2 the third-party package.  The failure behavior of this function will depend on the arguments passed to it - I think you want to refer to the `pattern` argument here instead of the RE2 package, because pattern is the main argument sent to `qio_regexp_create_compile` to compile into a regexp object (which is what is checked by `qio_regexp_ok(ret._regexp)`).  Does that make sense?

Comment (piyush0411) [reply modules/standard/Regexp.chpl:453]: @lydia-duncan Mam
The argument `pattern` should be either string or byte, as mentioned. So will it be ok to consider that the procedure throws an error if the argument is not of this type?

Comment (lydia-duncan) [reply modules/standard/Regexp.chpl:453]: Unfortunately, no.  The type of the pattern argument is restricted by the [where clause](https://chapel-lang.org/docs/latest/language/spec/procedures.html#where-clauses) - if the pattern argument doesn't match the types listed there, that will result in a compilation error because we don't have an alternate function to call.  I can definitely see how you got confused, though!  The throws keyword is right before the where clause syntactically, so I can see thinking they are related.  But no, what is thrown from a function is more tied to `throw` in the *body* of the function - the `throws` keyword in the declaration is merely to indicate that it can throw, not what it throws (which is why this documentation update is so important)

Comment (piyush0411) [reply modules/standard/Regexp.chpl:453]: ***
// Converts from Regexp error code to RE2 error code.
// Maybe some day they will diverge.  In any event, this
// hides the existence of Regexp from RE2 users.
static RE2::ErrorCode RegexpErrorToRE2(re2::RegexpStatusCode code) {
  switch (code) {
    case re2::kRegexpSuccess:
      return RE2::NoError;
    case re2::kRegexpInternalError:
      return RE2::ErrorInternal;
    case re2::kRegexpBadEscape:
      return RE2::ErrorBadEscape;
    case re2::kRegexpBadCharClass:
      return RE2::ErrorBadCharClass;
    case re2::kRegexpBadCharRange:
      return RE2::ErrorBadCharRange;
    case re2::kRegexpMissingBracket:
      return RE2::ErrorMissingBracket;
    case re2::kRegexpMissingParen:
      return RE2::ErrorMissingParen;
    case re2::kRegexpUnexpectedParen:
      return RE2::ErrorUnexpectedParen;
    case re2::kRegexpTrailingBackslash:
      return RE2::ErrorTrailingBackslash;
    case re2::kRegexpRepeatArgument:
      return RE2::ErrorRepeatArgument;
    case re2::kRegexpRepeatSize:
      return RE2::ErrorRepeatSize;
    case re2::kRegexpRepeatOp:
      return RE2::ErrorRepeatOp;
    case re2::kRegexpBadPerlOp:
      return RE2::ErrorBadPerlOp;
    case re2::kRegexpBadUTF8:
      return RE2::ErrorBadUTF8;
    case re2::kRegexpBadNamedCapture:
      return RE2::ErrorBadNamedCapture;
  }
  return RE2::ErrorInternal;
}
***
Are these error codes of some use? I think we can interpret some possible reasons for the error. 
But it might be difficult to write them down into the documentation.

Comment (piyush0411) [reply modules/standard/Regexp.chpl:453]: > I have read the modules for re2 and regexp, studied the function definition of the function we are working with.
An error is thrown if re2 && re2->ok() evaluates to false. In this file, in line 1022, the error statement is Invalid RE2. So should this be the reason for throwing an error?
In my opinion, this error is related to the improper creation of re2, as mentioned here

So going on with this documentation, I think that the function `ok()` returns `noError` if `re2` is created properly, else if it evaluates to `false`, one of the above error codes should be returned. So we should make use of these error codes in the comment.
What do you say about this?

Comment (lydia-duncan) [reply modules/standard/Regexp.chpl:453]: > Are these error codes of some use?

It looks like they are used by the RE2 implementation, but that the individual errors are converted into the message field for the Chapel-level `BadRegexpError` class.  So I don't think we need to go into them - if we move to having all those be individual Chapel Error subclasses, then the person making that change can update the documentation to cover them.

Comment (piyush0411) [reply modules/standard/Regexp.chpl:453]: I meant that we can use these errors to put some kind of constraints on our argument `pattern`. And then add it to the documentation.

Comment (lydia-duncan) [reply modules/standard/Regexp.chpl:453]: We could, though there are a lot of them so I would worry about the documentation feeling a bit bloated.  Especially because the module documentation contains a lot of information about how to specify the regular expressions already.  I'd lean towards not doing so for now and we can come back to it later if someone really wants it.

Comment (piyush0411) [reply modules/standard/Regexp.chpl:453]: @lydia-duncan 
Can you please tell me what could be a possible reason for this function to throw an error. I have gone through multiple files related to regular expression but couldn't find the proper explanation for this.

Comment (lydia-duncan) [reply modules/standard/Regexp.chpl:453]: You're most of the way there - I was worried listing the RE2 error codes would be too detailed, how would you generalize it?

Comment (piyush0411) [reply modules/standard/Regexp.chpl:453]: I am thinking of something like
```
/*
:throws BadRegexError: If the argument passed has syntactical errors.
Refer to RE2 Error Codes for more details.
*/
```
In this way, we can tell the user what could be the possible reason and where to look to solve that error.
Let me know what you think of this.

Comment (lydia-duncan) [reply modules/standard/Regexp.chpl:453]: I like it!  Is there a place with all the error codes that we can link to?  (Such as a link to a page from the RE2 implementation?)

Comment (piyush0411) [reply modules/standard/Regexp.chpl:453]: >static RE2::ErrorCode RegexpErrorToRE2(re2::RegexpStatusCode code) {
switch (code) {
case re2::kRegexpSuccess:
return RE2::NoError;
case re2::kRegexpInternalError:
return RE2::ErrorInternal;
case re2::kRegexpBadEscape:
return RE2::ErrorBadEscape;
case re2::kRegexpBadCharClass:
return RE2::ErrorBadCharClass;
case re2::kRegexpBadCharRange:
return RE2::ErrorBadCharRange;
case re2::kRegexpMissingBracket:
return RE2::ErrorMissingBracket;
case re2::kRegexpMissingParen:
return RE2::ErrorMissingParen;
case re2::kRegexpUnexpectedParen:
return RE2::ErrorUnexpectedParen;
case re2::kRegexpTrailingBackslash:
return RE2::ErrorTrailingBackslash;
case re2::kRegexpRepeatArgument:
return RE2::ErrorRepeatArgument;
case re2::kRegexpRepeatSize:
return RE2::ErrorRepeatSize;
case re2::kRegexpRepeatOp:
return RE2::ErrorRepeatOp;
case re2::kRegexpBadPerlOp:
return RE2::ErrorBadPerlOp;
case re2::kRegexpBadUTF8:
return RE2::ErrorBadUTF8;
case re2::kRegexpBadNamedCapture:
return RE2::ErrorBadNamedCapture;
}
return RE2::ErrorInternal;
}

These error codes are mentioned  [here](https://github.com/chapel-lang/chapel/blob/master/third-party/re2/re2-src/re2/re2.cc#L70) in Chapel Code. Can this be used for our purpose?

Comment (lydia-duncan) [reply modules/standard/Regexp.chpl:453]: I was hoping more for a webpage hosted by RE2

Comment (piyush0411) [reply modules/standard/Regexp.chpl:453]: Then we can refer to the `re2` repository on Github [here](https://github.com/google/re2/blob/master/re2/re2.h).

Comment (lydia-duncan) [reply modules/standard/Regexp.chpl:453]: Yeah, that looks fine to me

**Review by lydia-duncan — APPROVED** [review]
Awesome, thanks!  I think the built docs look good as well.  I'll merge this on Monday

**Pull request merged**
